### ISSUE TITLE
SynEditCodeFolding not for XE

### ISSUE THREAD

**Issue by geheimniswelten** [SynEdit/SynEdit#82]
> ```
> // Code Folding
> {$IFDEF SYN_DELPHI_2009_UP}
>   {$DEFINE SYN_CodeFolding}
> {$ENDIF}
> ```
Since D2009?

Unknown namespaces in Delphi XE.
Changed to and so it is OK:
```
uses
  Graphics,
  Types,
  Classes,
  SysUtils,
  Math,
  Generics.Defaults,
  Generics.Collections,
  SynEditHighlighter;
```

but then:
> SynEditCodeFolding.pas(355) Fehler: E2003 Undeklarierter Bezeichner: 'List'
> SynEditCodeFolding.pas(357) Fehler: E2003 Undeklarierter Bezeichner: 'FromLine'
> SynEditCodeFolding.pas(358) Fehler: E2003 Undeklarierter Bezeichner: 'ToLine'
> SynEditCodeFolding.pas(359) Fehler: E2003 Undeklarierter Bezeichner: 'Collapsed'
> SynEditCodeFolding.pas(359) Fehler: E2015 Operator ist auf diesen Operandentyp nicht anwendbar
> SynEditCodeFolding.pas(377) Fehler: E2003 Undeklarierter Bezeichner: 'List'
> SynEditCodeFolding.pas(378) Fehler: E2003 Undeklarierter Bezeichner: 'ToLine'
> SynEditCodeFolding.pas(384) Fehler: E2003 Undeklarierter Bezeichner: 'FromLine'
> SynEditCodeFolding.pas(408) Fehler: E2003 Undeklarierter Bezeichner: 'List'
> ...

**Comment by geheimniswelten** [comment]
OK, the whole project is no longer backwards compatible.  :(

SynEdit.pas:
System.Win.ComObj.OleCheck (HR); <- System.Win. is unknown

But from TRect.Width and TRect.Height I gave up first. (since XE3?)
Upgrading the third-party component in XE for the upcoming upgrade of the project to DX10.2.

**Comment by uschuster** [comment]
That the whole project might not compile with old versions is a temporary thing due to the latest changes. I try to keep it compilable even with Delphi 7, but right now I do not have (much) time. Worst case X-mas is not too far away.

I would prefer code folding for Delphi 2007 as well and looking at that code folding topic follows after making the project compilable again with all supported versions (Delphi 7 till 10.2 Tokyo).

**Comment by pyscripter** [comment]
The project should still be backward compatible.

The ComObj stuff is part of the effort by Christian Budde to replace GDI painting with Direct2D.  You can disable all this code by undefining the SYN_DirectWrite.   Since the code is experimental I think it should be undefined by default.

Similarly all Code folding stuff is inside IFDEF declarations:
```
{$IFDEF SYN_CodeFolding}
{$ENDIF}
```
The Code folding code uses generics and anonymous methods which I believe were introduced in Delphi 2009  (see for example [this](http://blog.marcocantu.com/blog/before_anonymous.html)).   However, I have only compiled the code with XE7 and 10.2, so I cannot confirm this.

I have not tested it but I believe that if you undefined both SYN_CodeFolding and SYN_DirectWrite the code should compile with earlier Delphi versions.

**Comment by uschuster** [comment]
I have fixed some compilation issues, disabled Code Folding for XE2 and lower, disabled DirectWrite for XE and lower and now the runtime packages compile with D2007 till 10.2.

Compilation with D7 requires reverting the UTF-8 file format of three files (so far I do not like the fix for #47) and it requires a small fix in SynEditMiscClasses.pas due to the use of TBitmap.SetSize.

**Comment by CWBudde** [comment]
As far as I have tested it, it seems to work in Delphi7 (except for the fact that you need to delete the BOM at the beginning of some units).
The fix mentioned in #47 is not optimal, especially regarding legibility, but it seems to work so far (I haven't done extensive tests). I will close this now as codefolding does work in XE now (though the demo needs some manual adaptions to make it compile)
